File: python-examples/13-parallel-review/workflows.py
"""Parallel review workflow examples.

Demonstrates how to run multiple workflows in parallel and aggregate results.
Useful for scenarios like:
- Multi-reviewer document review
- Parallel data processing
- Fan-out/fan-in patterns
"""

from pydantic import BaseModel

from polos import workflow, WorkflowContext
from polos.types.types import BatchWorkflowInput


class ReviewRequest(BaseModel):
    """Request for a single review."""

    reviewer_id: str
    document_id: str
    content: str


class ReviewResult(BaseModel):
    """Result from a single review."""

    reviewer_id: str
    document_id: str
    approved: bool
    score: int  # 1-10
    comments: str


class AggregatedReview(BaseModel):
    """Aggregated result from multiple reviews."""

    document_id: str
    total_reviews: int
    approved_count: int
    average_score: float
    all_approved: bool
    reviews: list[dict]


@workflow(id="single_review")
async def single_review(ctx: WorkflowContext, payload: ReviewRequest) -> ReviewResult:
    """Individual review workflow.

    This is invoked in parallel for each reviewer.
    """
    # Simulate review analysis
    result = await ctx.step.run(
        "analyze_document",
        analyze_document,
        payload.document_id,
        payload.content,
    )

    # Generate review decision
    review = await ctx.step.run(
        "generate_review",
        generate_review,
        payload.reviewer_id,
        result,
    )

    return ReviewResult(
        reviewer_id=payload.reviewer_id,
        document_id=payload.document_id,
        approved=review["approved"],
        score=review["score"],
        comments=review["comments"],
    )


def analyze_document(document_id: str, content: str) -> dict:
    """Analyze document content."""
    return {
        "document_id": document_id,
        "word_count": len(content.split()),
        "quality_score": 8,  # Simulated
    }


def generate_review(reviewer_id: str, analysis: dict) -> dict:
    """Generate a review based on analysis."""
    score = analysis.get("quality_score", 5)
    return {
        "approved": score >= 6,
        "score": score,
        "comments": f"Review by {reviewer_id}: Quality score {score}/10",
    }


@workflow(id="parallel_review")
async def parallel_review(ctx: WorkflowContext, payload: dict) -> AggregatedReview:
    """Run multiple reviews in parallel and aggregate results.

    Uses batch_invoke_and_wait to run all reviews concurrently.
    """
    document_id = payload.get("document_id", "doc-1")
    content = payload.get("content", "Sample document content")
    reviewers = payload.get("reviewers", ["alice", "bob", "charlie"])

    # Create batch of review requests
    review_requests = [
        BatchWorkflowInput(
            id="single_review",
            payload=ReviewRequest(
                reviewer_id=reviewer,
                document_id=document_id,
                content=content,
            ),
        )
        for reviewer in reviewers
    ]

    # Run all reviews in parallel and wait for all to complete
    results = await ctx.step.batch_invoke_and_wait(
        "parallel_reviews",
        review_requests,
    )

    # Aggregate results
    reviews = []
    approved_count = 0
    total_score = 0

    for batch_result in results:
        if batch_result.success and batch_result.result:
            review = batch_result.result
            reviews.append({
                "reviewer_id": review.reviewer_id,
                "approved": review.approved,
                "score": review.score,
                "comments": review.comments,
            })
            if review.approved:
                approved_count += 1
            total_score += review.score

    total_reviews = len(reviews)
    average_score = total_score / total_reviews if total_reviews > 0 else 0
    all_approved = approved_count == total_reviews

    return AggregatedReview(
        document_id=document_id,
        total_reviews=total_reviews,
        approved_count=approved_count,
        average_score=average_score,
        all_approved=all_approved,
        reviews=reviews,
    )


@workflow(id="data_chunk_processor")
async def data_chunk_processor(ctx: WorkflowContext, payload: dict) -> dict:
    """Process data in parallel chunks.

    Demonstrates fan-out/fan-in pattern for data processing.
    """
    data = payload.get("data", [])
    chunk_size = payload.get("chunk_size", 10)

    # Split data into chunks
    chunks = [data[i : i + chunk_size] for i in range(0, len(data), chunk_size)]

    # Create batch of chunk processing requests
    chunk_requests = [
        BatchWorkflowInput(
            id="process_chunk",
            payload={"chunk": chunk, "chunk_index": i},
        )
        for i, chunk in enumerate(chunks)
    ]

    # Process all chunks in parallel
    results = await ctx.step.batch_invoke_and_wait(
        "parallel_chunks",
        chunk_requests,
    )

    # Aggregate results
    all_processed = []
    for batch_result in results:
        if batch_result.success and batch_result.result:
            all_processed.extend(batch_result.result.get("processed", []))

    return {
        "total_items": len(data),
        "chunks_processed": len(results),
        "processed_items": len(all_processed),
        "results": all_processed,
    }


@workflow(id="process_chunk")
async def process_chunk(ctx: WorkflowContext, payload: dict) -> dict:
    """Process a single chunk of data."""
    chunk = payload.get("chunk", [])
    chunk_index = payload.get("chunk_index", 0)

    processed = await ctx.step.run(
        f"process_items_{chunk_index}",
        lambda items: [item.upper() if isinstance(item, str) else item * 2 for item in items],
        chunk,
    )

    return {
        "chunk_index": chunk_index,
        "processed": processed,
    }


@workflow(id="fire_and_forget_batch")
async def fire_and_forget_batch(ctx: WorkflowContext, payload: dict) -> dict:
    """Launch multiple workflows without waiting for results.

    Uses batch_invoke to start workflows and return immediately.
    """
    tasks = payload.get("tasks", [])

    # Create batch of task requests
    task_requests = [
        BatchWorkflowInput(
            id="background_task",
            payload={"task_id": task.get("id"), "data": task.get("data")},
        )
        for task in tasks
    ]

    # Launch all tasks in parallel (non-blocking)
    handles = await ctx.step.batch_invoke(
        "launch_background_tasks",
        task_requests,
    )

    # Return execution IDs for tracking
    return {
        "launched": len(handles),
        "execution_ids": [h.id for h in handles],
    }


@workflow(id="background_task")
async def background_task(ctx: WorkflowContext, payload: dict) -> dict:
    """Background task for fire-and-forget pattern."""
    task_id = payload.get("task_id")
    data = payload.get("data", {})

    result = await ctx.step.run(
        "execute_task",
        lambda: {"task_id": task_id, "status": "completed"},
    )

    return result

File: typescript-examples/13-parallel-review/workflows.ts
/**
 * Parallel review workflow examples.
 *
 * Demonstrates how to run multiple workflows in parallel and aggregate results.
 * Useful for scenarios like:
 * - Multi-reviewer document review
 * - Parallel data processing
 * - Fan-out/fan-in patterns
 */

import { defineWorkflow } from '@polos/sdk';
import type { BatchWorkflowInput, BatchStepResult } from '@polos/sdk';

// ============================================================================
// Payload / Result Types
// ============================================================================

export interface ReviewRequest {
  reviewerId: string;
  documentId: string;
  content: string;
}

export interface ReviewResult {
  reviewerId: string;
  documentId: string;
  approved: boolean;
  score: number; // 1-10
  comments: string;
}

interface AggregatedReview {
  documentId: string;
  totalReviews: number;
  approvedCount: number;
  averageScore: number;
  allApproved: boolean;
  reviews: { reviewerId: string; approved: boolean; score: number; comments: string }[];
}

interface ChunkProcessorPayload {
  data: unknown[];
  chunkSize: number;
}

interface ChunkProcessorResult {
  totalItems: number;
  chunksProcessed: number;
  processedItems: number;
  results: unknown[];
}

interface ChunkPayload {
  chunk: unknown[];
  chunkIndex: number;
}

interface ChunkResult {
  chunkIndex: number;
  processed: unknown[];
}

interface FireAndForgetPayload {
  tasks: { id: string; data: Record<string, unknown> }[];
}

interface BackgroundTaskPayload {
  taskId: string;
  data: Record<string, unknown>;
}

// ============================================================================
// Helper Functions
// ============================================================================

function analyzeDocument(documentId: string, content: string): Record<string, unknown> {
  return {
    documentId,
    wordCount: content.split(/\s+/).length,
    qualityScore: 8, // Simulated
  };
}

function generateReview(reviewerId: string, analysis: Record<string, unknown>): Record<string, unknown> {
  const score = (analysis['qualityScore'] as number) ?? 5;
  return {
    approved: score >= 6,
    score,
    comments: `Review by ${reviewerId}: Quality score ${String(score)}/10`,
  };
}

// ============================================================================
// Single Review Workflow
// ============================================================================

export const singleReview = defineWorkflow<ReviewRequest, unknown, ReviewResult>(
  { id: 'single_review' },
  async (ctx, payload) => {
    const result = await ctx.step.run(
      'analyze_document',
      () => analyzeDocument(payload.documentId, payload.content),
    );

    const review = await ctx.step.run(
      'generate_review',
      () => generateReview(payload.reviewerId, result),
    );

    return {
      reviewerId: payload.reviewerId,
      documentId: payload.documentId,
      approved: review['approved'] as boolean,
      score: review['score'] as number,
      comments: review['comments'] as string,
    };
  },
);

// ============================================================================
// Parallel Review Workflow
// ============================================================================

export const parallelReview = defineWorkflow<Record<string, unknown>, unknown, AggregatedReview>(
  { id: 'parallel_review' },
  async (ctx, payload) => {
    const documentId = (payload['documentId'] as string) ?? 'doc-1';
    const content = (payload['content'] as string) ?? 'Sample document content';
    const reviewers = (payload['reviewers'] as string[]) ?? ['alice', 'bob', 'charlie'];

    // Create batch of review requests
    const reviewRequests: BatchWorkflowInput[] = reviewers.map((reviewer) => ({
      workflow: 'single_review',
      payload: {
        reviewerId: reviewer,
        documentId,
        content,
      } satisfies ReviewRequest,
    }));

    // Run all reviews in parallel and wait for all to complete
    const results = await ctx.step.batchInvokeAndWait<ReviewResult>(
      'parallel_reviews',
      reviewRequests,
    );

    // Aggregate results (each entry is a BatchStepResult wrapper)
    const reviews: AggregatedReview['reviews'] = [];
    let approvedCount = 0;
    let totalScore = 0;

    for (const entry of results) {
      const review = entry.result;
      if (!review) continue;
      reviews.push({
        reviewerId: review.reviewerId,
        approved: review.approved,
        score: review.score,
        comments: review.comments,
      });
      if (review.approved) {
        approvedCount += 1;
      }
      totalScore += review.score;
    }

    const totalReviews = reviews.length;
    const averageScore = totalReviews > 0 ? totalScore / totalReviews : 0;
    const allApproved = approvedCount === totalReviews;

    return {
      documentId,
      totalReviews,
      approvedCount,
      averageScore,
      allApproved,
      reviews,
    };
  },
);

// ============================================================================
// Data Chunk Processor (fan-out/fan-in)
// ============================================================================

export const dataChunkProcessor = defineWorkflow<ChunkProcessorPayload, unknown, ChunkProcessorResult>(
  { id: 'data_chunk_processor' },
  async (ctx, payload) => {
    const data = payload.data;
    const chunkSize = payload.chunkSize;

    // Split data into chunks
    const chunks: unknown[][] = [];
    for (let i = 0; i < data.length; i += chunkSize) {
      chunks.push(data.slice(i, i + chunkSize));
    }

    // Create batch of chunk processing requests
    const chunkRequests: BatchWorkflowInput[] = chunks.map((chunk, i) => ({
      workflow: 'process_chunk',
      payload: { chunk, chunkIndex: i } satisfies ChunkPayload,
    }));

    // Process all chunks in parallel
    const results = await ctx.step.batchInvokeAndWait<ChunkResult>(
      'parallel_chunks',
      chunkRequests,
    );

    // Aggregate results (each entry is a BatchStepResult wrapper)
    const allProcessed: unknown[] = [];
    for (const entry of results) {
      if (entry.result) {
        allProcessed.push(...entry.result.processed);
      }
    }

    return {
      totalItems: data.length,
      chunksProcessed: results.length,
      processedItems: allProcessed.length,
      results: allProcessed,
    };
  },
);

// ============================================================================
// Process Chunk (invoked by data_chunk_processor)
// ============================================================================

export const processChunk = defineWorkflow<ChunkPayload, unknown, ChunkResult>(
  { id: 'process_chunk' },
  async (ctx, payload) => {
    const chunk = payload.chunk;
    const chunkIndex = payload.chunkIndex;

    const processed = await ctx.step.run(
      `process_items_${String(chunkIndex)}`,
      () =>
        chunk.map((item) =>
          typeof item === 'string' ? item.toUpperCase() : (item as number) * 2,
        ),
    );

    return { chunkIndex, processed };
  },
);

// ============================================================================
// Fire and Forget Batch
// ============================================================================

export const fireAndForgetBatch = defineWorkflow<FireAndForgetPayload, unknown, Record<string, unknown>>(
  { id: 'fire_and_forget_batch' },
  async (ctx, payload) => {
    const tasks = payload.tasks;

    // Create batch of task requests
    const taskRequests: BatchWorkflowInput[] = tasks.map((task) => ({
      workflow: 'background_task',
      payload: { taskId: task.id, data: task.data } satisfies BackgroundTaskPayload,
    }));

    // Launch all tasks in parallel (non-blocking)
    const handles = await ctx.step.batchInvoke(
      'launch_background_tasks',
      taskRequests,
    );

    return {
      launched: handles.length,
      executionIds: handles.map((h) => h.executionId),
    };
  },
);

// ============================================================================
// Background Task (invoked by fire_and_forget_batch)
// ============================================================================

export const backgroundTask = defineWorkflow<BackgroundTaskPayload, unknown, Record<string, unknown>>(
  { id: 'background_task' },
  async (ctx, payload) => {
    const result = await ctx.step.run(
      'execute_task',
      () => ({ taskId: payload.taskId, status: 'completed' }),
    );

    return result;
  },
);
